vendor/bam/src/json/header.rs

```diff
@@ -0,0 +1,240 @@
+use serde::{
+    de::{DeserializeOwned, Deserializer},
+    Deserialize, Serialize,
+};
+use std::collections::{BTreeMap, HashMap};
+
+fn deserialize_compact_value<'de, D>(deserializer: D) -> Result<serde_json::Value, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    match serde_json::Value::deserialize(deserializer)? {
+        serde_json::Value::Object(_) => Err(serde::de::Error::custom(
+            "header value must not be an object in compact form",
+        )),
+        serde_json::Value::Null => Err(serde::de::Error::custom(
+            "header value must not be null in compact form",
+        )),
+        other => Ok(other),
+    }
+}
+
+fn deserialize_extended_value<'de, D>(deserializer: D) -> Result<serde_json::Value, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    match serde_json::Value::deserialize(deserializer)? {
+        serde_json::Value::Null => Err(serde::de::Error::custom(
+            "header value must not be null in extended form",
+        )),
+        other => Ok(other),
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(untagged)]
+enum CompactOrExtended {
+    Extended {
+        #[serde(deserialize_with = "deserialize_extended_value")]
+        value: serde_json::Value,
+        #[serde(default)]
+        parameters: BTreeMap<String, serde_json::Value>,
+    },
+    #[serde(deserialize_with = "deserialize_compact_value")]
+    Compact(serde_json::Value),
+}
+
+impl CompactOrExtended {
+    fn value(&self) -> serde_json::Value {
+        match self {
+            CompactOrExtended::Compact(value) | CompactOrExtended::Extended { value, .. } => {
+                value.clone()
+            }
+        }
+    }
+
+    fn take_parameter(&mut self, key: &'static str) -> Option<serde_json::Value> {
+        match self {
+            CompactOrExtended::Compact { .. } => None,
+            CompactOrExtended::Extended { parameters, .. } => parameters.remove(key),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(transparent)]
+pub struct Header {
+    inner: CompactOrExtended,
+}
+
+impl Header {
+    pub fn value<V: DeserializeOwned>(&self) -> Result<V, serde_json::Error> {
+        serde_json::from_value(self.inner.value())
+    }
+
+    /// Returns the parameter with the provided key converted into the type `P`.
+    ///
+    /// If the parameter doesn't exist, `Null` is passed to the conversion
+    /// (which will very likely fail). TODO: Is this a good idea? :D
```

Comment:
so, you are saying that you know that by returning `Null` it will fail somewhere else? Why not just returning an error right here?